### PR TITLE
Add close() methods throughout the SecureJar->JarContents chain to allow closing the backing file systems

### PR DESCRIPTION
**Pull request by shartte**
**BREAKING CHANGE**

### CODE REVIEW

File: src/main/java/cpw/mods/niofs/union/UnionFileSystem.java [hunk 1]
```diff
@@ -159,8 +159,11 @@ public UnionFileSystemProvider provider() {
     }
 
     @Override
-    public void close() {
+    public void close() throws IOException {
         provider().removeFileSystem(this);
+        for (var embeddedFs : embeddedFileSystems.values()) {
```
Comment (Technici4n) [inline src/main/java/cpw/mods/niofs/union/UnionFileSystem.java:164]: Technically if there is an exception here, the remaining file systems won't get closed.

Comment (marchermans) [inline src/main/java/cpw/mods/niofs/union/UnionFileSystem.java:164]: Please surround the underlying close calls with a try-catch and continue closing the remaining File systems.

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/cpw/mods/niofs/union/UnionFileSystem.java (hunk 1 above)
Comment (shartte) [reply src/main/java/cpw/mods/niofs/union/UnionFileSystem.java:161]: Addressed

Comment (shartte) [reply src/main/java/cpw/mods/niofs/union/UnionFileSystem.java:161]: Addressed

File: src/main/java/cpw/mods/jarhandling/SecureJar.java [hunk 2]
```diff
@@ -86,6 +82,12 @@ static SecureJar from(JarContents contents, JarMetadata metadata) {
      */
     Path getRootPath();
 
+    /**
+     * Closes the underlying file system resources (if any).
+     * Renders this object unusable.
+     */
+    void close() throws IOException;
```
Comment (Matyrobbrt) [inline src/main/java/cpw/mods/jarhandling/SecureJar.java:89]: You can extend Closeable instead.

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/cpw/mods/jarhandling/SecureJar.java (hunk 2 above)
Comment (shartte) [reply src/main/java/cpw/mods/jarhandling/SecureJar.java:89]: ... that just reintroduces the implementation of AutoClosable

public interface Closeable
extends [AutoCloseable](https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html)

And that seriously disables the warning?

Comment (shartte) [reply src/main/java/cpw/mods/jarhandling/SecureJar.java:89]: Will leave it like this

**Review by Matyrobbrt — APPROVED** [review]

**Review by marchermans — APPROVED** [review]

**Pull request merged**